generate-geodatabase-replica-from-feature-service/src/main/java/com/esri/arcgismaps/sample/generategeodatabasereplicafromfeatureservice/MainActivity.kt

```diff
@@ -0,0 +1,256 @@
+    private val generateButton by lazy {
+        activityMainBinding.generateButton
+    }
+
+    // shows the geodatabase loading progress
+    private val progressDialog by lazy {
+        GenerateGeodatabaseDialogLayoutBinding.inflate(layoutInflater)
+    }
+
+    // local file path to the geodatabase
+    private val geodatabaseFilePath by lazy {
+        getExternalFilesDir(null)?.path + getString(R.string.portland_trees_geodatabase_file)
+    }
+
+    // a red boundary line showing the filtered map extents for the features
+    private val boundarySymbol = SimpleLineSymbol(SimpleLineSymbolStyle.Solid, Color.red, 5f)
+
+    // creates a graphic overlay
+    private val graphicsOverlay: GraphicsOverlay = GraphicsOverlay()
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+
+        // authentication with an API key or named user is
+        // required to access basemaps and other location services
+        ArcGISEnvironment.apiKey = ApiKey.create(BuildConfig.API_KEY)
+        lifecycle.addObserver(mapView)
+
+        // create and add a map with a Topographic basemap style
+        val map = ArcGISMap(BasemapStyle.ArcGISTopographic)
+        // set the max map extents to that of the feature service
+        // representing portland area
+        map.maxExtent = Envelope(
+            -13687689.2185849, 5687273.88331375,
+            -13622795.3756647, 5727520.22085841,
+            spatialReference = SpatialReference.webMercator()
+        )
+        // configure mapview assignments
+        mapView.apply {
+            this.map = map
+            // add the graphics overlay to display the boundary
+            graphicsOverlays.add(graphicsOverlay)
+        }
+
+        lifecycleScope.launch {
+            // show the error and return if map load failed
+            map.load().onFailure {
+                showError("Unable to load map")
+                return@launch
+            }
+            // create a geodatabase sync task with the feature service url
+            // This feature service shows a web map of portland street trees,
+            // their attributes, as well as related inspection information
+            val geodatabaseSyncTask = GeodatabaseSyncTask(getString(R.string.feature_server_url))
+            geodatabaseSyncTask.load().onFailure {
+                // if the metadata load fails, show the error and return
+                showError("Failed to fetch geodatabase metadata")
+                return@launch
+            }
+            // enable the generate button and add an onClickListener
+            generateButton.isEnabled = true
+            generateButton.setOnClickListener {
+                mapView.visibleArea?.let { polygon ->
+                    // start the geodatabase generation process
+                    generateGeodatabase(geodatabaseSyncTask, map, polygon.extent)
+                }
+            }
+        }
```

Comment:
Not sure I quite understand the first part. From my understanding, since `load()` is a suspending call, the scope will pause and wait for it to return. Using `getOrElse`, `onFailure` are just ways to access the value of `Result<T>` that `load()` has returned. [kotlin.Result](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-result/).

I've been using `getOrElse` to get the value of `Result<T>` in cases where it is something other than `Unit/Any`, otherwise using `onFailure`. Especially to avoid nested `onSuccess()/onFailure()`statements.

You're right about the second part. In fact, that's how I had written it initially. But I coupled the creation of `geodatabaseSyncTask` and `generateButton.setOnClickListener{}` with the successful loads as their scope exits in case of failures. Since there is no _refresh_, the only way to retry is to re-launch the app. I'm happy to refactor this though, if the former is the accepted pattern.

Reply:
Seems like your understanding is correct here. We can use your approach here as the standard approach. 👍 

With the second part, the `generateButton` is disabled by default and is enabled when `geodatabaseSyncTask` is loaded. This way the button cannot be clicked until the task successfully loads. So I feel like it could be coupled outside of the coroutine scope:
```kotlin
// create a geodatabase sync task with the feature service url
// This feature service shows a web map of portland street trees,
// their attributes, as well as related inspection information
val geodatabaseSyncTask = GeodatabaseSyncTask(getString(R.string.feature_server_url))

// set the button click listener
generateButton.setOnClickListener {
    mapView.visibleArea?.let { polygon ->
        // start the geodatabase generation process
        generateGeodatabase(geodatabaseSyncTask, map, polygon.extent)
    }
}

lifecycleScope.launch {
    // show the error and return if map load failed
    map.load().onFailure {
        showError("Unable to load map")
        return@launch
    }
    geodatabaseSyncTask.load().onFailure {
        // if the metadata load fails, show the error and return
        showError("Failed to fetch geodatabase metadata")
        return@launch
    }
    // task is loaded, enable the generate button
    generateButton.isEnabled = true
}
```